### PR TITLE
Turn off text sync if we don't _really_ need it

### PR DESCRIPTION
**Pull request by davidwengier**
At the moment the always active language server only does Ctrl+Q, and pull diagnostics sometimes, but is opting in to getting a message about every text edit just in case. Seems like overkill, and showed up on a perf trace.

### CODE REVIEW

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/AlwaysActivateInProcLanguageClient.cs [hunk 1]
```diff
@@ -46,17 +46,24 @@ public override string Name
...
+            var isPullDiagnostics = this.Workspace.IsPullDiagnostics(InternalDiagnosticsOptions.NormalDiagnosticMode);
+            return new VSServerCapabilities
             {
                 TextDocumentSync = new TextDocumentSyncOptions
                 {
-                    Change = TextDocumentSyncKind.Incremental,
-                    OpenClose = true,
+                    // If pull diagnostics isn't enabled, then we don't really need text sync so can save a bunch of
+                    // requests that the platform would otherwise make, and state to store all of the text documents.
+                    // Strictly speaking workspace symbols could be not 100% accurate as a result, but the trade off
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/LanguageClient/AlwaysActivateInProcLanguageClient.cs:57]: can you clarify what you mean by: Strictly speaking workspace symbols could be not 100% accurate as a result

?

**Comment by CyrusNajmabadi** [comment]
16.9 candidate.  Tagging @jinujoseph

File: src/VisualStudio/Core/Def/Implementation/LanguageClient/AlwaysActivateInProcLanguageClient.cs (hunk 1 above)
Comment (davidwengier) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/AlwaysActivateInProcLanguageClient.cs:57]: Just that by turning off text sync, you get back the problems that it solves, namely that the LSP requests will work on whatever text is in the buffer, rather that being ordered correctly with each individual edit.

For local editing this wouldn't be a problem, but for codespaces in theory the LSP request for a search could come in before the buffer has been updated by liveshare. In practice as far as I know its impossible, and would only manifest if someone were to type in a member name and then immediately use VS Search to try to find it.

Comment (CyrusNajmabadi) [reply src/VisualStudio/Core/Def/Implementation/LanguageClient/AlwaysActivateInProcLanguageClient.cs:57]: I see. So it opens up a small period of time when you can get stale results, but things eventually catch up?

**Comment by CyrusNajmabadi** [comment]
@NTaylorMullen Can You validate that this doesn't break you at all?

**Comment by NTaylorMullen** [comment]
> @NTaylorMullen Can You validate that this doesn't break you at all?

Doesn't break Razor since we always have "pull diagnostics" on.

That being said, I feel like this is setting you all up for failure down the line. Once the LSP client refactor comes into play their entire synchronization model is based off of the text doc sync messages. Also, in a world where Roslyn goes OOP (I know a lot further out) having a TextBuffer to operate on wouldn't be a safe assumption would it?

**Comment by davidwengier** [comment]
> I feel like this is setting you all up for failure down the line

100%, this is most definitely a temporary mitigation. When this server gets used for real for just about anything else, this will have to be reverted.

**Comment by NTaylorMullen** [comment]
> 100%, this is most definitely a temporary mitigation. When this server gets used for real for just about anything else, this will have to be reverted.

If that's the case the client refactor effort isn't that far away (month max)

**Comment by CyrusNajmabadi** [comment]
That's a month of grumpy cyrus.  This was found because my VS is just gettin crushed with allocations, and this was a heavy hitter there.  I don't want to spend a month paying an expensive price for LSP when doing local development when i'm not using a single LSP feature.  :-/  

Plzfixkthx

**Comment by NTaylorMullen** [comment]
> That's a month of grumpy cyrus. This was found because my VS is just gettin crushed with allocations, and this was a heavy hitter there. I don't want to spend a month paying an expensive price for LSP when doing local development when i'm not using a single LSP feature. :-/
> 
> Plzfixkthx

Aside from turning off text sync is there another path (I don't have a lot of background on the actual issue)? Asking because we may have a month of grumpy cyrus but if we don't solve it now we'll have a year+ of grumpy cyrus once they make the transition to the new client infrastructure

**Comment by CyrusNajmabadi** [comment]
We def need to figure out why simple text change notifications are leading to such a high cost in our side. @davidwengier to continue investigation there :-)

**Comment by davidwengier** [comment]
Sam has fix that apparently saves 10% of allocations which I'm curious to see. Fundamentally by its nature though, text sync needs to store the text of every open document, so allocations are expected.

I suspect there are better Roslyn workspace or solution APIs that could be used (or created) for constructing the LSP solution though, so its worth discussing.

**Comment by dibarbet** [comment]
Just wanted to check, do we have an issue tracking the underlying problem?  We should definitely fix it as I presume any 16.9 razor project with lots of files will hit the same issue cyrus is hitting, since we do text sync there.

**Review by dibarbet — APPROVED** [review]
I'm fine with doing this as long as we get this reenabled for 16.10p1.  an lsp world with non lsp text sync is confusing

**Comment by davidwengier** [comment]
> Just wanted to check, do we have an issue tracking the underlying problem?

Not as far as i'm aware. There is a teams chat waiting for you when you get back :)

**Review by CyrusNajmabadi — APPROVED** [review]

**Comment by davidwengier** [comment]
Discussed offline, we're not going to take this but will follow up with https://github.com/dotnet/roslyn/issues/50875 and other perf fixes instead.

**Pull request closed without merging**